### PR TITLE
HGCAL Electronics mapping (v2)

### PR DESCRIPTION
**Pull request by pfs**
This PR supersedes #43751. It is rebased on top of `CMSSW_14_1_0_pre2` and it's made of a single commit which squashes all the ones in #43751. Below the description and the validation of #43751 are copied and updated

#### PR description:

Adding the infrastructure needed to

- provide the electronics to detector identifier mapping
- provide the dense indexer for unpacked data and conditions (made available through EventSetup)
- provide detailed information on the cell and modules (made available through EventSetup in SoAs)

The tool has been discussed in 

- HGCAL DPG [slides](https://indico.cern.ch/event/1358621/) and provisional documentation in [google doc](https://docs.google.com/document/d/1kGKsMBQ8CWC7kJN71aJmsUv4vVDJ8ZlG_D1ColCEuAk/edit#bookmark=id.8ys9dfg99vqv) 
- AlCaDB [slides 1](https://indico.cern.ch/event/1372780/) [slides 2](https://indico.cern.ch/event/1386295/) 
- #43751

It relies on the mappings in https://github.com/cms-data/Geometry-HGCalMapping/pull/1

The requests to leave EVENTSETUP_RECORD_REG and COND_SERIALIZABLE commented are not implemented as they generate compilation errors and break the testers. The HGCAL DPG is however aware that the move to CondDB is to occur in a future time-scale (end of this year/begin on next year). 

#### PR validation:

This PR is not expected to impact any workflow thus we expect 100% success in the tests. The PR is however accompanied by a series of unit tests and analyzers (under the test) sub-directories. They have been used to validate the content produced by the new pieces of code being added to the repository.

@kerstinlovisa @jalimena

### CODE REVIEW

**Comment by pfs** [comment]
please test with https://github.com/cms-data/Geometry-HGCalMapping/pull/1

**Comment by jfernan2** [comment]
type hgcal

**Comment by pfs** [comment]
please test with https://github.com/cms-data/Geometry-HGCalMapping/pull/1

File: CondFormats/HGCalObjects/src/classes.h [hunk 1]
```diff
@@ -0,0 +1 @@
+#include "CondFormats/HGCalObjects/src/headers.h"
```
[the remark this answers is not in the document]
Comment (pfs) [reply CondFormats/HGCalObjects/src/classes.h:1]: thanks - indeed we followed strictly as recommended in https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCondObjectsTutorial . Other variations tried resulted in compilation errors

File: DataFormats/HGCalDigi/interface/HGCalElectronicsId.h [hunk 2]
```diff
@@ -39,27 +42,31 @@ class HGCalElectronicsId {
   */
   HGCalElectronicsId() : value_(0) {}
   explicit HGCalElectronicsId(
```
[the remark this answers is not in the document]
Comment (pfs) [reply DataFormats/HGCalDigi/interface/HGCalElectronicsId.h:44]: Indeed, in the previous PR Ihave made such mistake. `explicit` has been introduced in PR #44484

File: Geometry/HGCalMapping/test/BuildFile.xml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<bin file="testHGCalMapFileParser.cc">
+  <flags TEST_RUNNER_ARGS=" Geometry/HGCalMapping/data/ModuleMaps/modulelocator_CEminus_V15p5.txt"/>
```
[the remark this answers is not in the document]
Comment (pfs) [reply Geometry/HGCalMapping/test/BuildFile.xml:2]: yes, this test unit has been added and forces https://github.com/cms-data/Geometry-HGCalMapping/pull/1 to be used explicitly

File: CondFormats/HGCalObjects/src/classes.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (makortel) [reply CondFormats/HGCalObjects/src/classes.h:1]: What I found out in https://github.com/cms-sw/cmssw/pull/43751#discussion_r1498121290 is that `scram b` has a specific "CondFormat serialization" phase that requires the `headers.h` to exist. I'd suggest to document that requirement better :)

**Comment by pfs** [comment]
No need to duplicate - conflicts solved in https://github.com/cms-sw/cmssw/pull/43751.

**Pull request closed without merging**